src/core/server/http/http_service.ts

```diff
@@ -44,10 +56,37 @@ export type HttpServiceSetup = Omit<HttpServerSetup, 'registerRouter'> & {
    * // handler is called when '${my-plugin-id}/path' resource is requested with `GET` method
    * router.get({ path: '/path', validate: false }, (context, req, res) => res.ok({ content: 'ok' }));
    * ```
+   * @public
+   */
+  createRouter: (path: string, plugin?: PluginOpaqueId) => IRouter;
+  /**
+   * Register a context provider for a route handler.
+   * @example
+   * ```ts
+   *  // my-plugin.ts
+   *  deps.http.registerRouteHandlerContext(
+   *    'myApp',
+   *    (context, req) => {
+   *     async function search (id: string) {
+   *       return await context.elasticsearch.adminClient.callAsInternalUser('endpoint', id);
+   *     }
+   *     return { search };
+   *    }
+   *  );
    *
-   * @internal
-   * */
-  createRouter: (path: string) => IRouter;
+   * // my-route-handler.ts
+   *  router.get({ path: '/', validate: false }, async (context, req, res) => {
+   *    const response = await context.myApp.search(...);
+   *    return res.ok(response);
+   *  });
+   * ```
+   * @public
+   */
+  registerRouteHandlerContext: (
+    pluginOpaqueId: PluginOpaqueId,
+    contextName: RequestHandlerContextNames,
+    provider: RequestHandlerContextProvider
+  ) => RequestHandlerContextContainer;
```

Comment:
This won't enforce type safety as much as possible. You'll want to use a type parameter for the specific contextName here to make sure that the `provider` returns the correct type given the concrete `contextName`:

```
  registerRouteHandlerContext<T extends RequestHandlerContextNames>: (
    pluginOpaqueId: PluginOpaqueId,
    contextName: T,
    provider: IContextProvider<RequestHandlerContext, T>
  ) => RequestHandlerContextContainer;
```